### ISSUE TITLE
Add `get_leaf_depth` method to `MerkleStore`

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/crypto#119]
This method is needed to support MASM implementation of Tiered STM. The signature of the method could be:
```Rust
pub fn get_leaf_depth(&self, root: Word, index: u64) -> Result<u8, MerkleError>
```
This method would traverse the tree specified by the `root` from the top and return as soon as it finds a leaf node (i.e., node with no children) or a root of an empty subtree.

The error conditions would be:
- There is no tree for the specified `root` in the store.
- We got to depth 64 but the node at that depth is not a leaf.

**Comment by vlopes11** [comment]
To disambiguate, I renamed `index` to `path` so we don't mix with the `depth/index` pair

**Comment by vlopes11** [comment]
There is a odd behavior with the store (or maybe I'm just overlooking something). I think this test shouldn't pass

```rust
#[test]
fn node_path_should_be_truncated() {
    let keys: &[u64] = &[
        0b11010010_11001100_11001100_11001100_11001100_11001100_11001100_11001100,
        0b00000000_11010010_11001100_11001100_11001100_11001100_11001100_11001101,
    ];

    let mut store = MerkleStore::new();
    let root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();

    // insert first node - works as expected
    let key = keys[0];
    let depth = 64;
    let node = [Felt::new(key); WORD_SIZE];
    let index = NodeIndex::new(depth, key);
    let root = store.set_node(root, index, node).unwrap().root;
    let result = store.get_node(root, index).unwrap();
    let path = store.get_path(root, index).unwrap().path;
    assert_eq!(node, result);
    assert!(path.verify(index.value(), result, &root));

    // insert second node on a different depth - works as expected
    let key = keys[1];
    let depth = 56;
    let node = [Felt::new(key); WORD_SIZE];
    let index = NodeIndex::new(depth, key);
    let root = store.set_node(root, index, node).unwrap().root;
    let result = store.get_node(root, index).unwrap();
    let path = store.get_path(root, index).unwrap().path;
    assert_eq!(node, result);
    assert!(path.verify(index.value(), result, &root));

    // attempt to fetch a path of the second node to depth 64
    // note: there is no such pre-image, this command shouldn't pass as the path will fail!
    let index = NodeIndex::new(64, key);
    let result = store.get_node(root, index).unwrap();
    let path = store.get_path(root, index).unwrap().path;
    assert!(path.verify(index.value(), result, &root));
}
```

I also probably need to update this

https://github.com/0xPolygonMiden/crypto/blob/9389f2fb40bce30b5429a7d3820416f1f8e723ed/src/merkle/store/mod.rs#L314

to

```rust
let mut index = NodeIndex::new(path.depth() as u8, index_value);
```

**Comment by bobbinth** [comment]
Closed by #125.
